backend/service/topology/querybuilder.go

```diff
@@ -0,0 +1,141 @@
+package topology
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+	"strconv"
+	"strings"
+
+	sq "github.com/Masterminds/squirrel"
+
+	topologyv1 "github.com/lyft/clutch/backend/api/topology/v1"
+)
+
+const (
+	column            = "column."
+	metadata          = "metadata."
+	queryDefaultLimit = 100
+	maxResultLimit    = 1000
+)
+
+func paginatedQueryBuilder(
+	filter *topologyv1.SearchRequest_Filter,
+	sort *topologyv1.SearchRequest_Sort,
+	pageToken string,
+	limit int,
+) (sq.SelectBuilder, error) {
+	queryLimit := queryDefaultLimit
+	if limit > maxResultLimit {
+		return sq.SelectBuilder{}, errors.New("Maximum query limit is 1000")
+	} else if limit > 0 {
+		queryLimit = limit
+	}
+
+	pageNum, err := strconv.Atoi(pageToken)
+	if err != nil {
+		return sq.SelectBuilder{}, err
+	}
+
+	queryOffset := 0
+	if pageNum > 0 {
+		queryOffset = pageNum * limit
+	}
+
+	query := sq.StatementBuilder.PlaceholderFormat(sq.Dollar).
+		Select("id", "data", "metadata").
+		From("topology_cache").
+		Limit(uint64(queryLimit)).
+		Offset(uint64(queryOffset))
+
+	if filter != nil {
+		query, err = filterQueryBuilder(query, filter)
+		if err != nil {
+			return sq.SelectBuilder{}, err
+		}
+	}
+
+	if sort != nil {
+		query = sortQueryBuilder(query, sort)
+	}
+
+	return query, nil
+}
+
+func filterQueryBuilder(query sq.SelectBuilder, f *topologyv1.SearchRequest_Filter) (sq.SelectBuilder, error) {
+	if f == nil {
+		return query, nil
+	}
+
+	if f.Search != nil && len(f.Search.Field) > 0 {
+		if strings.HasPrefix(f.Search.Field, column) {
+			query = query.Where(sq.Expr("quote_literal(?) LIKE ?", sq.Expr(strings.TrimPrefix(f.Search.Field, column)), fmt.Sprintf("%%%s%%", f.Search.Text)))
+		} else if strings.HasPrefix(f.Search.Field, metadata) {
+			mdQuery := convertMetadataToQuery(strings.TrimPrefix(f.Search.Field, metadata))
+			query = query.Where(sq.Expr("quote_literal(?) LIKE ?", sq.Expr(mdQuery), fmt.Sprintf("%%%s%%", f.Search.Text)))
+		}
+	}
+
+	if len(f.TypeUrl) > 0 {
+		query = query.Where(sq.Eq{"resolver_type_url": f.TypeUrl})
+	}
+
+	if f.Metadata != nil {
+		metadataJson, err := json.Marshal(f.Metadata)
+		if err != nil {
+			return sq.SelectBuilder{}, err
+		}
+		query = query.Where(sq.Expr("metadata @> ?::jsonb", metadataJson))
+	}
+
+	return query, nil
+}
+
+func sortQueryBuilder(query sq.SelectBuilder, s *topologyv1.SearchRequest_Sort) sq.SelectBuilder {
+	if s == nil {
+		return query
+	}
+
+	if s.Direction > topologyv1.SearchRequest_Sort_UNSPECIFIED && len(s.Field) > 0 {
```

Comment:
`if s.Direction > topologyv1.SearchRequest_Sort_UNSPECIFIED`
i thought this would default to ASC anyways? otherwise this validation belongs in the proto

Reply:
Ah yeah i can delete this